core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -41,15 +45,25 @@ public Object getFieldValue(Metric o) {
     }
   };
 
-  private static final IndexDefinition<Metric> HOSTNAME_INDEX = new IndexDefinition<Metric>(false) {
-    @Override
-    public Object getFieldValue(Metric o) {
-      return o.getHostname();
-    }
-  };
+  private static final IndexDefinition<Metric> HOSTNAME_ID_INDEX =
+      new IndexDefinition<Metric>(false) {
+        @Override
+        public Object getFieldValue(Metric o) {
+          return getHostnameAndId(o.getHostname(), o.getInstanceId());
+        }
+      };
+
+  private static String getHostnameAndId(String hostname, String id) {
+    String str = hostname == null ? "" : hostname;
+    str = str.replace('.', '_');
+    str += id == null ? "" : ":" + id;
+    return str;
+  }
 
   private final IndexedSet<Metric> mWorkerMetrics =
-      new IndexedSet<>(FULL_NAME_INDEX, NAME_INDEX, HOSTNAME_INDEX);
+      new IndexedSet<>(FULL_NAME_INDEX, NAME_INDEX, HOSTNAME_ID_INDEX);
```

Comment:
this is confusing since worker metrics don't include ID. Can we just have two different indexes, one for hostname, one for hostname+id?

Reply:
hmm I feel it's fine as hostname only is a special form of hostname+id.

Reply:
this feels like an unnecessary hack to save a few lines of boilerplate code. The data type <hostname> is different from the type <hostname, id>, not a special case of it.

Reply:
or perhaps we could get rid of the concept of hostname, and just have a concept of "ID". Then the ID for workers would be their hostname, while the ID for clients would be hostname + a client specific ID

Reply:
having the `ID` sounds good to me.